sr_robot_launch/src/sr_robot_launch/mock_sr_ur_robot_hw.py

```diff
@@ -0,0 +1,252 @@
+
+    def restart_safety(self):
+        self.set_safety_mode('NORMAL')
+        self.set_robot_mode('POWER_OFF')
+        self.set_program_running(False)
+
+    def power_on(self):
+        self.set_safety_mode('NORMAL')
+        self.set_robot_mode('IDLE')
+        self.set_program_running(False)
+
+    def brake_release(self):
+        self.set_safety_mode('NORMAL')
+        self.set_robot_mode('RUNNING')
+        self.set_program_running(False)
+
+    def unlock_protective_stop(self):
+        self.set_safety_mode('NORMAL')
+        self.set_robot_mode('RUNNING')
+        self.set_program_running(False)
+
+    def power_off(self):
+        self.set_safety_mode('NORMAL')
+        self.set_robot_mode('POWER_OFF')
+        self.set_program_running(False)
+
+    def protective_stop(self):
+        self.set_safety_mode('PROTECTIVE_STOP')
+        self.set_robot_mode('RUNNING')
+        self.set_program_running(False)
+
+    def emergency_stop(self):
+        self.set_safety_mode('ROBOT_EMERGENCY_STOP')
+        self.set_robot_mode('IDLE')
+        self.set_program_running(False)
+
+    def resend_robot_program(self):
+        if (self.get_safety_mode().safety_mode.mode == SafetyMode.NORMAL and
+                self.get_robot_mode().robot_mode.mode == RobotMode.RUNNING):
+            self.set_program_running(True)
+        else:
+            self.set_program_running(False)
+
+    def fault(self):
+        self.set_safety_mode('FAULT')
+        self.set_robot_mode('STOPPED')
+        self.set_program_running(False)
+
+    def update_state(self, state):
+        if state not in self.states:
+            rospy.logwarn("Invalid state: %s valid states are: %s", state, str(self.states))
+            return
+        if state == 'ROBOT_STOPPED':
+            self.set_safety_mode('NORMAL')
+            self.set_robot_mode('POWER_OFF')
+            self.set_program_running(False)
+        if state == 'ROBOT_IDLE':
+            self.set_safety_mode('NORMAL')
+            self.set_robot_mode('IDLE')
+            self.set_program_running(False)
+        if state == 'ROBOT_RUNNING':
+            self.set_safety_mode('NORMAL')
+            self.set_robot_mode('RUNNING')
+            self.set_program_running(False)
+        if state == 'PROTECTIVE_STOP':
+            self.set_safety_mode('PROTECTIVE_STOP')
+            self.set_robot_mode('RUNNING')
+            self.set_program_running(False)
+        if state == 'EMERGENCY-STOP':
+            self.set_safety_mode('ROBOT_EMERGENCY_STOP')
+            self.set_robot_mode('IDLE')
+            self.set_program_running(False)
+        if state == 'PROGRAM_RUNNING':
+            self.set_safety_mode('NORMAL')
+            self.set_robot_mode('RUNNING')
+            self.set_program_running(True)
+
+
+class MockUrRobotHW(object):
+    def __init__(self, side='right'):
+        if 'left' not in side and 'right' not in side:
+            rospy.logerr("side: %s not valid. Valid sides are 'left, 'right'", side)
```

Comment:
please use "new style" formatting
https://realpython.com/python-string-formatting/

Reply:
even for strings formatted by rospy?

Reply:
hm, is there a reason not to?